substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -0,0 +1,1527 @@
+						.block_body(h.hash)
+						.await
+						.unwrap_or_else(|e| {
+							log::warn!(target: LOG_TARGET, "Compute ready light transactions: error request: {}", e);
+							None
+						})
+						.unwrap_or_default()
+						.into_iter()
+						.map(|t| api.hash_and_length(&t).0);
+					all_extrinsics.extend(extrinsics);
+				}
+
+				let before_count = tmp_view.pool.validated_pool().status().ready;
+				let in_pool_tags = tmp_view.pool.validated_pool().extrinsics_tags(&all_extrinsics);
+
+				let mut tags = Vec::new();
+				for i in in_pool_tags {
+					match i {
+						// reuse the tags for extrinsics that were found in the pool
+						Some(t) => tags.extend(t),
+						None => {},
+					}
+				}
+				let _ = tmp_view.pool.validated_pool().prune_tags(tags);
+
+				let after_count = tmp_view.pool.validated_pool().status().ready;
+				log::info!(target: LOG_TARGET,
+					"fatp::ready_at_light {} from {} before: {} to be removed: {} after: {} took:{:?}",
+					at,
+					best_view.at.hash,
+					before_count,
+					all_extrinsics.len(),
+					after_count,
+					start.elapsed()
+				);
+				Box::new(tmp_view.pool.validated_pool().ready())
+			} else {
+				let empty: ReadyIteratorFor<ChainApi> = Box::new(std::iter::empty());
+				log::info!(target: LOG_TARGET, "fatp::ready_at_light {} -> empty, took:{:?}", at, start.elapsed());
+				empty
+			}
+		})
+	}
+
+	/// Waits for the set of ready transactions for a given block up to a specified timeout.
+	///
+	/// This method combines two futures:
+	/// - The `ready_at` future, which waits for the ready transactions resulting from the full
+	/// maintenance process to be available.
+	/// - The `ready_at_light` future, used as a fallback if the timeout expires before `ready_at`
+	/// completes. This provides a best-effort, ready set of transactions as a result light
+	/// maintain.
+	///
+	/// Returns a future resolving to a ready iterator of transactions.
+	fn ready_at_with_timeout_internal(
+		&self,
+		at: Block::Hash,
+		timeout: std::time::Duration,
+	) -> PolledIterator<ChainApi> {
+		log::info!(target: LOG_TARGET, "fatp::ready_at_with_timeout at {:?} allowed delay: {:?}", at, timeout);
+
+		let timeout = futures_timer::Delay::new(timeout);
+		let fall_back_ready = self.ready_at_light(at).map(|ready| Some(ready));
+		let ready_at = self.ready_at(at);
+
+		let maybe_ready = async move {
+			select! {
+				ready = ready_at => Some(ready),
+				_ = timeout => {
+					log::warn!(target: LOG_TARGET,
+						"Timeout fired waiting for transaction pool at block: ({:?}). \
+						Proceeding with production.",
+						at,
+					);
+					None
+				}
+			}
+		};
+
+		//todo: if ready_at is immediately available we could skip waiting for fallback.
+
+		Box::pin(async {
+			let (maybe_ready, fall_back_ready) =
+				futures::future::join(maybe_ready.boxed(), fall_back_ready.boxed()).await;
+			maybe_ready
+				.unwrap_or_else(|| fall_back_ready.expect("Fallback value is always Some. qed"))
+		})
```

Comment:
Spent a while on this, but did not find out how it can be done.
We need `light` to be run in parallel with `timeout` and `full`. Something like: `light` and (`timeout` or `full`).

Leaving here a playground for experiments: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b9d1fa9ecb10d5501bb5fe61641a669a